src/openms/source/CHEMISTRY/AASequence.cpp

```diff
@@ -107,6 +107,68 @@ namespace OpenMS
     return tmp;
   }
 
+  String AASequence::toUniModString() const
+  {
+    const AASequence & seq = *this;
+
+    String bs;
+    int unimod = -1;
+
+    if (seq.empty()) return bs;
+
+    if (seq.hasNTerminalModification())
+    {
+      const ResidueModification& mod = *(seq.getNTerminalModification());
+      unimod = mod.getUniModRecordId();
+      if (unimod > -1)
+      {
+        bs += ".(UniMod:" + String(unimod) + ")";
+      }
+      else
+      {
+        bs += ".[" + String(mod.getDiffMonoMass()) + "]";
+      }
+    }
+
+    for (Size i = 0; i != seq.size(); ++i)
+    {
+      const Residue& r = seq[i];
+      const String aa = r.getOneLetterCode() != "" ? r.getOneLetterCode() : "X";
+      if (r.isModified())
+      {
+        const ResidueModification& mod = *(r.getModification());
+        unimod = mod.getUniModRecordId();
+        if (unimod > -1)
+        {
+          bs += aa + "(UniMod:" + String(unimod) + ")";
+        }
+        else
+        {
+          bs += aa + "[" + String(r.getMonoWeight(Residue::Internal)) + "]";
+        }
+      }
+      else  // amino acid not modified
+      {
+        bs += aa;
+      }
+    }
+
+    if (seq.hasCTerminalModification())
+    {
+      const ResidueModification& mod = *(seq.getCTerminalModification());
+      unimod = mod.getUniModRecordId();
```

Comment:
this code is duplicated somewhat 3x...
IMHO it would make sense to write a private function `string getUnimod(const ResidueModification& mod, const char site) const` which assembles the partial string (`site` would be either `.` or the one-letter code).

Reply:
There are slight differences between the three parts (getDiffMonoMass vs getMonoWeight) and I'm thus not sure whether this would simplify the code.

Reply:
ok, you'll need a third parameter (weight), but this cuts down 50 lines of code into ~5.
Up to you of course :)